File: cloud/admin/src/routes/combos.js
// /api/combos — fallback-chain CRUD.

import { query, NotFoundError, ValidationError } from '@9router-cloud/shared';
import { readJson, ok, noContent } from '../lib/http.js';
import { requireSession, requireRole } from '../middleware/sessionAuth.js';

export async function listCombos(req, res) {
  const s = requireSession(req);
  const { rows } = await query(`
    SELECT id, slug, name, nodes, enabled, created_at, updated_at
    FROM combos
    WHERE tenant_id = $1
    ORDER BY slug ASC
  `, [s.tenantId]);
  ok(res, { items: rows.map(serialize) });
}

export async function createCombo(req, res) {
  const s = requireSession(req);
  requireRole(s, 'owner', 'admin');
  const body = await readJson(req);
  const slug = required(body, 'slug').toLowerCase().trim();
  if (!/^[a-z0-9_-]{1,40}$/.test(slug)) {
    throw new ValidationError('slug must be 1-40 chars of [a-z0-9_-]');
  }
  const nodes = body.nodes;
  validateNodes(nodes);
  const name = (body.name || slug).toString();
  const enabled = body.enabled !== false;

  let row;
  try {
    const r = await query(`
      INSERT INTO combos (tenant_id, slug, name, nodes, enabled)
      VALUES ($1, $2, $3, $4, $5)
      RETURNING id, slug, name, nodes, enabled, created_at, updated_at
    `, [s.tenantId, slug, name, JSON.stringify(nodes), enabled]);
    row = r.rows[0];
  } catch (e) {
    if (e.code === '23505') throw new ValidationError(`Combo ${slug} already exists`);
    throw e;
  }
  ok(res, serialize(row), 201);
}

export async function updateCombo(req, res, { params }) {
  const s = requireSession(req);
  requireRole(s, 'owner', 'admin');
  const slug = params.slug;
  const body = await readJson(req);
  const sets = [];
  const args = [];
  let i = 1;
  if (body.name !== undefined) { sets.push(`name = $${i++}`); args.push(String(body.name)); }
  if (body.nodes !== undefined) {
    validateNodes(body.nodes);
    sets.push(`nodes = $${i++}`); args.push(JSON.stringify(body.nodes));
  }
  if (body.enabled !== undefined) { sets.push(`enabled = $${i++}`); args.push(!!body.enabled); }
  if (sets.length === 0) throw new ValidationError('No updatable fields');
  sets.push(`updated_at = now()`);
  args.push(slug, s.tenantId);
  const { rows } = await query(
    `UPDATE combos SET ${sets.join(', ')}
     WHERE slug = $${i++} AND tenant_id = $${i}
     RETURNING id, slug, name, nodes, enabled, created_at, updated_at`,
    args,
  );
  if (rows.length === 0) throw new NotFoundError(`Combo ${slug} not found`);
  ok(res, serialize(rows[0]));
}

export async function deleteCombo(req, res, { params }) {
  const s = requireSession(req);
  requireRole(s, 'owner', 'admin');
  const slug = params.slug;
  const { rowCount } = await query(
    `DELETE FROM combos WHERE slug = $1 AND tenant_id = $2`,
    [slug, s.tenantId]
  );
  if (rowCount === 0) throw new NotFoundError(`Combo ${slug} not found`);
  noContent(res);
}

function validateNodes(nodes) {
  if (!Array.isArray(nodes) || nodes.length === 0) {
    throw new ValidationError('nodes must be a non-empty array');
  }
  for (let i = 0; i < nodes.length; i++) {
    const n = nodes[i];
    if (!n || typeof n !== 'object') throw new ValidationError(`nodes[${i}] must be an object`);
    if (!n.provider || typeof n.provider !== 'string') throw new ValidationError(`nodes[${i}].provider required`);
    if (!n.model || typeof n.model !== 'string') throw new ValidationError(`nodes[${i}].model required`);
  }
}

function required(obj, field) {
  if (obj[field] === undefined || obj[field] === null || obj[field] === '') {
    throw new ValidationError(`Missing field: ${field}`);
  }
  return String(obj[field]);
}

function serialize(row) {
  return {
    id: Number(row.id),
    slug: row.slug,
    name: row.name,
    nodes: row.nodes,
    enabled: row.enabled,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

File: internal/admin/combos.go
package admin

import (
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"github.com/amaozhao/lazirouter/internal/db"
	"github.com/amaozhao/lazirouter/internal/errs"
	"github.com/amaozhao/lazirouter/internal/httpx"
	"github.com/jackc/pgx/v5"
)

type comboDTO struct {
	ID        int64           `json:"id"`
	Slug      string          `json:"slug"`
	Name      string          `json:"name"`
	Nodes     json.RawMessage `json:"nodes"`
	Enabled   bool            `json:"enabled"`
	CreatedAt time.Time       `json:"createdAt"`
	UpdatedAt time.Time       `json:"updatedAt"`
}

func (d *Deps) ListCombos(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	rows, err := db.Pool().Query(r.Context(),
		`SELECT id, slug, name, nodes, enabled, created_at, updated_at FROM combos WHERE tenant_id = $1 ORDER BY id`, s.TenantID)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	defer rows.Close()
	items := []comboDTO{}
	for rows.Next() {
		var c comboDTO
		if err := rows.Scan(&c.ID, &c.Slug, &c.Name, &c.Nodes, &c.Enabled, &c.CreatedAt, &c.UpdatedAt); err != nil {
			httpx.WriteError(w, err)
			return
		}
		items = append(items, c)
	}
	httpx.WriteJSON(w, http.StatusOK, map[string]any{"items": items})
}

type createComboReq struct {
	Slug  string           `json:"slug"`
	Name  string           `json:"name"`
	Nodes []map[string]any `json:"nodes"`
}

func (d *Deps) CreateCombo(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	var req createComboReq
	if err := httpx.ReadJSON(r, &req); err != nil {
		httpx.WriteError(w, err)
		return
	}
	if req.Slug == "" || len(req.Nodes) == 0 {
		httpx.WriteError(w, errs.Validation("slug + nodes[] required"))
		return
	}
	if req.Name == "" {
		req.Name = req.Slug
	}
	var c comboDTO
	err = db.Pool().QueryRow(r.Context(), `
		INSERT INTO combos (tenant_id, slug, name, nodes)
		VALUES ($1, $2, $3, $4)
		RETURNING id, slug, name, nodes, enabled, created_at, updated_at
	`, s.TenantID, req.Slug, req.Name, marshalJSON(req.Nodes)).Scan(
		&c.ID, &c.Slug, &c.Name, &c.Nodes, &c.Enabled, &c.CreatedAt, &c.UpdatedAt,
	)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	httpx.WriteJSON(w, http.StatusOK, c)
}

type updateComboReq struct {
	Name    *string          `json:"name"`
	Nodes   []map[string]any `json:"nodes"`
	Enabled *bool            `json:"enabled"`
}

func (d *Deps) UpdateCombo(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	slug := r.PathValue("slug")
	if slug == "" {
		httpx.WriteError(w, errs.Validation("slug path param required"))
		return
	}
	var req updateComboReq
	if err := httpx.ReadJSON(r, &req); err != nil {
		httpx.WriteError(w, err)
		return
	}
	sets := []string{}
	args := []any{}
	if req.Name != nil {
		args = append(args, *req.Name)
		sets = append(sets, "name = $"+itoa(len(args)))
	}
	if req.Nodes != nil {
		args = append(args, marshalJSON(req.Nodes))
		sets = append(sets, "nodes = $"+itoa(len(args)))
	}
	if req.Enabled != nil {
		args = append(args, *req.Enabled)
		sets = append(sets, "enabled = $"+itoa(len(args)))
	}
	if len(sets) == 0 {
		httpx.WriteError(w, errs.Validation("no updatable fields"))
		return
	}
	args = append(args, slug, s.TenantID)
	q := `UPDATE combos SET ` + join(sets, ", ") + `, updated_at = now()
	      WHERE slug = $` + itoa(len(args)-1) + ` AND tenant_id = $` + itoa(len(args)) + `
	      RETURNING id, slug, name, nodes, enabled, created_at, updated_at`
	var c comboDTO
	err = db.Pool().QueryRow(r.Context(), q, args...).Scan(
		&c.ID, &c.Slug, &c.Name, &c.Nodes, &c.Enabled, &c.CreatedAt, &c.UpdatedAt,
	)
	if errors.Is(err, pgx.ErrNoRows) {
		httpx.WriteError(w, errs.NotFound("Combo not found"))
		return
	}
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	httpx.WriteJSON(w, http.StatusOK, c)
}

func (d *Deps) DeleteCombo(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	slug := r.PathValue("slug")
	if slug == "" {
		httpx.WriteError(w, errs.Validation("slug path param required"))
		return
	}
	tag, err := db.Pool().Exec(r.Context(),
		`DELETE FROM combos WHERE slug = $1 AND tenant_id = $2`, slug, s.TenantID)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	if tag.RowsAffected() == 0 {
		httpx.WriteError(w, errs.NotFound("Combo not found"))
		return
	}
	httpx.WriteJSON(w, http.StatusOK, map[string]any{"ok": true, "slug": slug})
}
